src/main/java/io/cryostat/jmcagent/S3ProbeTemplateService.java

```diff
@@ -0,0 +1,242 @@
+        var req = GetObjectRequest.builder().bucket(bucket).key(name).build();
+        return storage.getObject(req);
+    }
+
+    @Blocking
+    private List<S3Object> getObjects() {
+        var builder = ListObjectsV2Request.builder().bucket(bucket);
+        return storage.listObjectsV2(builder.build()).contents();
+    }
+
+    @Blocking
+    private ProbeTemplate convertObject(S3Object object) throws Exception {
+        var req = GetObjectTaggingRequest.builder().bucket(bucket).key(object.key()).build();
+        var tagging = storage.getObjectTagging(req);
+        var list = tagging.tagSet();
+        if (!tagging.hasTagSet() || list.isEmpty()) {
+            throw new Exception("No metadata found");
+        }
+        var decodedList = new ArrayList<Pair<String, String>>();
+        list.forEach(
+                t -> {
+                    var encodedKey = t.key();
+                    var decodedKey =
+                            new String(base64Url.decode(encodedKey), StandardCharsets.UTF_8).trim();
+                    var encodedValue = t.value();
+                    var decodedValue =
+                            new String(base64Url.decode(encodedValue), StandardCharsets.UTF_8)
+                                    .trim();
+                    decodedList.add(Pair.of(decodedKey, decodedValue));
+                });
+        var fileName =
+                decodedList.stream()
+                        .filter(t -> t.getKey().equals("fileName"))
+                        .map(Pair::getValue)
+                        .findFirst()
+                        .orElseThrow();
+        var classPrefix =
+                decodedList.stream()
+                        .filter(t -> t.getKey().equals("classPrefix"))
+                        .map(Pair::getValue)
+                        .findFirst()
+                        .orElseThrow();
+        var allowToString =
+                decodedList.stream()
+                        .filter(t -> t.getKey().equals("allowToString"))
+                        .map(Pair::getValue)
+                        .findFirst()
+                        .orElseThrow();
+        var allowConverter =
+                decodedList.stream()
+                        .filter(t -> t.getKey().equals("allowConverter"))
+                        .map(Pair::getValue)
+                        .findFirst()
+                        .orElseThrow();
+        // filename, classprefix, allowtostring, allowconverter
+        return new ProbeTemplate(
+                fileName,
+                classPrefix,
+                Boolean.valueOf(allowToString),
+                Boolean.valueOf(allowConverter));
+    }
+
+    @Blocking
+    public ProbeTemplate addTemplate(InputStream stream, String fileName)
+            throws IOException, SAXException {
+        try (stream) {
+            ProbeTemplate template = new ProbeTemplate();
+            template.setFileName(fileName);
+            template.deserialize(stream);
```

Comment:
IIRC the `deserialize()` call is just used here to ensure that the document conforms to the expected XML schema, right?

Either way, I see that `deserialize()` is called both here as well as in the `postProbe` endpoint method. It should be sufficient to just perform the validation once, here within the Service implementation.